doc executor description (#24)
exasim-project/NeoFOAM: doc/api/executor.rst

.. _api_executor:

Executor
========

Overview
^^^^^^^^

NeoFOAM uses the MPI+X approach for parallelism, where X is the execution space used for parallelism. The `Executor` class uses Kokkos and provides an interface for memory management and specifics were to execute the operations:

- `CPUExecutor`: run on the CPU with MPI
- `OMPExecutor`: run on the CPU with OpenMP and MPI
- `GPUExecutor`: run on the GPU with MPI

Design
^^^^^^

> One of the design goals was the ability to quickly switch between the executor models at runtime as already discussed in the field implementation:

Reviewer: ```suggestion
One of the design goals is the ability to quickly switch between the executor models at runtime.
```

.. code-block:: cpp

        NeoFOAM::GPUExecutor gpuExec {};
        NeoFOAM::CPUExecutor cpuExec {};

        NeoFOAM::Field<NeoFOAM::scalar> GPUField(gpuExec, 10);
        NeoFOAM::Field<NeoFOAM::scalar> CPUField(cpuExec, 10);

The `Executor` is a std::variant

.. code-block:: cpp

    using executor = std::variant<OMPExecutor, GPUExecutor, CPUExecutor>;

> and allows to switch between the different strategy for memory allocation and execution at runtime. We use `std::visit` to switch between the different strategies:

Reviewer: ```suggestion
and allows to switch between the different strategies for memory allocation and execution at runtime. We use `std::visit` to switch between the different strategies:
```
May be link to https://en.cppreference.com/w/cpp/utility/variant/visit
